uswds/uswds
# USWDS - Input Mask: (Error Feedback) Code Refactor

# Summary
**Improve invalid character feedback with error message.** Add error messages (customizable via data attributes) to give proper user feedback when an invalid character is not added to the input value.

## Breaking change

This is not a breaking change however there two minor warnings to note:
* This update adds an error message string below input masks which could result in a small shift in content.
* For sites with translated content: this update adds error message strings in english only.


## Related issue

Closes #5481. Closes #5482.

## Related Pull Request
[USWDS-Site: Add changelog for input mask [#6203]](https://github.com/uswds/uswds-site/pull/3001)

## Preview link
[Preview link](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/ml-input-mask-refactor/?path=/story/components-form-inputs-text-input-mask--ssn)

## Problem statement
Previously, the input mask component provided minimal visual feedback and no auditory feedback when users attempted to enter an invalid character. Usability testing revealed that users were often unaware that their input was rejected. The goal was to provide clear, accessible feedback through error messages and ARIA attributes, ensuring users understood why their input was rejected and what was expected instead.

## Solution
This solution introduced descriptive, dynamic error messages that informed users about invalid inputs and clarified what was expected, whether a letter or a number. Additionally, enhanced ARIA labeling ensured screen readers conveyed the same feedback, with customizable messages that could provide further guidance if needed.

## Testing and review
Navigate to text input mask component (alphanumeric is best for testing all the error message types at once but the others should still be tested for regression)

1. Tab into input, no error should display
2. Type “A” displays no error
3. Type “AB” displays “**You must enter a number”**
4. Type “A1” displays no error
5. Type “A12” displays “**You must enter a letter”**
6. Type “A1B” displays no error
7.  Finishing input with “A1B 2C3” displays no error
8. Continuing to type displays error “**You have reached the maximum number of characters allowed”**
9. BACKSPACE clears the error
10. ARROW LEFT and ARROW RIGHT do not trigger an error
11. CMD + BACKSPACE clears the input with no error displayed
12. Pasting “A1B2C3” formats text and displays no error 
*Paste tests should only be done with keyboard, issues with paste click events will be covered in a seperate issue
14. Pasting “A1B2222” displays “**Some pasted characters were not accepted.”**
15. Using navigational keys do not display an error:  HOME, END, ARROW LEFT, ARROW RIGHT, PAGE UP, PAGE DOWN, TAB
16. When testing using a screen reader the error message should be read to user 1000ms after the error message displays. This message should be able to be customized separately from the one displayed visually. 

*When error message does not display it should also be hidden with no visual padding or styling
*There should be no regressions or changes to mask functionality outside of partial improvements to paste functionality. This is only an error message enhancement.
